server/test/endpoints/role.spec.js

```diff
@@ -0,0 +1,369 @@
+    (done) => {
+      const invalidNewRole = SpecHelper.generateRandomRole('d0');
+      invalidNewRole.id = 1;
+      client.post('/roles')
+      .send(invalidNewRole)
+      .set({ 'xsrf-token': adminUser.token })
+      .end((error, response) => {
+        console.log('rorle', response.body)
+        expect(response.status).to.equal(400);
+        done();
+      });
+    });
+
+    xit(`should NOT allow an Admin user with VALID token create a
+     DUPLICATE Role`,
+    (done) => {
+      client.post('/roles')
+      .set({ 'x-access-token': adminUser.token })
+      .send(newRole)
+      .end((error, response) => {
+        expect(response.status).to.equal(400);
+        done();
+      });
+    });
+
+    xit(`should NOT allow an Admin user with VALID token create an
+     Invalid Role`,
+    (done) => {
+      client.post('/roles')
+      .set({ 'x-access-token': adminUser.token })
+      .send({})
+      .end((error, response) => {
+        expect(response.status).to.equal(400);
+        done();
+      });
+    });
+
+    xit(`should NOT allow any User (Admin, Regular...) with an INVALID token
+    create a Role`,
+    (done) => {
+      const testRole = SpecHelper.generateRandomRole('learner');
+      client.post('/roles')
+      .set({ 'x-access-token': 'invalid token' })
+      .send(testRole)
+      .end((error, response) => {
+        expect(response.status).to.equal(401);
+        done();
+      });
+    });
+
+    xit(`should NOT allow an authenticated NON-Admin User with a VALID token
+    create a Role`,
+    (done) => {
+      const testRole = SpecHelper.generateRandomRole('new learner');
+      client.post('/roles')
+      .set({ 'x-access-token': regularUser.token })
+      .send(testRole)
+      .end((error, response) => {
+        expect(response.status).to.equal(403);
+        done();
+      });
+    });
+  });
+
+  xdescribe('Update Role', () => {
+    const newRole = SpecHelper.generateRandomRole('update');
+    before((done) => {
+      client.post('/roles')
+      .send(newRole)
+      .set({ 'x-access-token': adminUser.token })
+      .end((error, response) => {
+        // set our new role id
+        newRole.id = response.body.id;
+        done();
+      });
+    });
+
+    it('should allow only an Admin user with VALID token UPDATE a Role',
+    (done) => {
+      const newTitle = 'new title';
+      client.put(`/roles/${newRole.id}`)
+      .set({ 'x-access-token': adminUser.token })
+      .send({ title: newTitle })
+      .end((error, response) => {
+        expect(response.status).to.equal(200);
+        done();
+      });
+    });
+
+    it(`should NOT allow an Admin user with VALID token UPDATE a
+    NON-EXISTENT Role`, (done) => {
+      const newTitle = 'new title1';
+      client.put(`/roles/${300}`)
+      .set({ 'x-access-token': adminUser.token })
+      .send({ title: newTitle })
+      .end((error, response) => {
+        expect(response.status).to.equal(404);
```

Comment:
Great job writing so many tests for your routes. Try to avoid checking only status codes in your test suits, also assert that the response is what was expected.

I understand that the app is still being worked on but try to add tests for the models and middlewares used in the application. For the front-end, try to add tests for all your components, reducers and action creators. Also add end-to-end tests for user activities like ‘login’, ‘create document’ etc.